File: backend/app/repositories/profiles.py
from __future__ import annotations
from typing import Any

from postgrest import APIError
from supabase import Client

from app.db.supabase import get_supabase_client

PROFILE_TABLE = "users_profile"

class ProfilesRepository:
    def __init__(self, client: Client | None = None) -> None:
        self.client = client or get_supabase_client()

    async def get_by_id(self, user_id: str) -> dict[str, Any] | None:
        response = (
            self.client.table(PROFILE_TABLE)
            .select("*")
            .eq("id", user_id)
            .limit(1)
            .execute()
        )

        data = getattr(response, "data", None) or []
        return data[0] if data else None

    async def create(self, user_id: str, attributes: dict[str, Any]) -> dict[str, Any]:
        payload: dict[str, Any] = {"id": user_id, **attributes}

        try:
            response = self.client.table(PROFILE_TABLE).insert(payload).execute()
        except APIError as exc:
            if getattr(exc, "code", None) == "23505":
                raise
            raise

        data = getattr(response, "data", None)
        if data and len(data) > 0:
            return data[0]

        # Some Supabase Python client versions return no data; fetch freshly.
        created = await self.get_by_id(user_id)
        if not created:
            raise RuntimeError("Profile insertion succeeded but record could not be retrieved")
        return created

    async def update(self, user_id: str, updates: dict[str, Any]) -> dict[str, Any]:
        response = (
            self.client.table(PROFILE_TABLE)
            .update(updates)
            .eq("id", user_id)
            .select("*")
            .single()
            .execute()
        )
        return response.data

File: src/lib/database/profiles.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { UserProfile } from '../../types';
import type { Database } from './types';
import { tryGetSupabaseClient } from './client';

export async function fetchProfileById(userId: string): Promise<UserProfile | null> {
  const client = tryGetSupabaseClient();
  if (!client) return null;

  const { data, error } = await (client.from('users_profile') as any)
    .select('*')
    .eq('id', userId)
    .maybeSingle();

  if (error) {
    console.error('[Supabase] Failed to fetch profile', error);
    return null;
  }

  return (data as UserProfile | null) ?? null;
}

export async function createProfile(initialProfile: Database['public']['Tables']['users_profile']['Insert']): Promise<UserProfile | null> {
  const client = tryGetSupabaseClient();
  if (!client) return null;

  const { data, error } = await (client.from('users_profile') as any)
    .insert([initialProfile])
    .select()
    .single();

  if (error) {
    console.error('[Supabase] Failed to create profile', error);
    return null;
  }

  return data as UserProfile;
}

export async function updateProfileById(
  userId: string,
  updates: Database['public']['Tables']['users_profile']['Update']
): Promise<UserProfile | null> {
  const client = tryGetSupabaseClient();
  if (!client) return null;

  const { data, error } = await (client.from('users_profile') as any)
    .update(updates)
    .eq('id', userId)
    .select()
    .single();

  if (error) {
    console.error('[Supabase] Failed to update profile', error);
    return null;
  }

  return data as UserProfile;
}
